src/dataplane-protocol/src/smartstream.rs

```diff
@@ -0,0 +1,167 @@
+pub use encoding::{
+    SmartStreamRuntimeError, SmartStreamRuntimeErrorBuilder, SmartStreamType, SmartStreamOutput,
+};
+
+pub type Result<T> = std::result::Result<T, SmartStreamRuntimeErrorBuilder>;
+
+mod encoding {
+    use std::fmt;
+    use crate::Offset;
+    use crate::record::Record;
+    use fluvio_protocol::{Encoder, Decoder};
+
+    /// A type used to capture and serialize errors from within a SmartStream
+    #[derive(thiserror::Error, Debug, Default, Clone, PartialEq, Encoder, Decoder)]
+    pub struct SmartStreamRuntimeError {
+        /// Error hint: meant for users, not for code
+        pub hint: String,
+        /// The offset of the Record that had a runtime error
+        pub offset: i64,
+        /// The type of SmartStream that had a runtime error
+        pub kind: SmartStreamType,
+        /// A preview of the Record key that caused this error
+        pub record_key: String,
+        /// A preview of the Record value that caused this error
+        pub record_value: String,
```

Comment:
Should be RecordData

Reply:
This isn't intended to be a lossless representation of the Record, but a user-readable repr

Reply:
let's not worry about user readability here.